lib/skate/detours/route_segments.ex

```diff
@@ -21,49 +21,69 @@ defmodule Skate.Detours.RouteSegments do
-      ...>   Util.Location.new(0, 1),
-      ...>   Util.Location.new(0, 2)
+      ...>   [
+      ...>     Location.new(0.0, 0.0),
+      ...>     Location.new(0.0, 1.0),
+      ...>     Location.new(0.0, 2.0),
+      ...>     Location.new(0.0, 3.0)
+      ...>   ],
+      ...>   Location.new(0.0, 1.3),
+      ...>   Location.new(0.0, 2.6)
       ...> )
       {:ok,
         %Skate.Detours.RouteSegments.Result{
-          before_detour: [%Util.Location{latitude: 0, longitude: 0}, %Util.Location{latitude: 0, longitude: 1}],
-          detour: [%Util.Location{latitude: 0, longitude: 1}, %Util.Location{latitude: 0, longitude: 2}],
-          after_detour: [%Util.Location{latitude: 0, longitude: 2}, %Util.Location{latitude: 0, longitude: 3}]
+          before_detour: [
+            %Util.Location{latitude: 0.0, longitude: 0.0},
+            %Util.Location{latitude: 0.0, longitude: 1.0},
+            %Util.Location{latitude: 0.0, longitude: 1.3}
+          ],
+          detour: [
+            %Util.Location{latitude: 0.0, longitude: 1.3},
+            %Util.Location{latitude: 0.0, longitude: 2.0},
+            %Util.Location{latitude: 0.0, longitude: 2.6}
+          ],
+          after_detour: [
+            %Util.Location{latitude: 0.0, longitude: 2.6},
+            %Util.Location{latitude: 0.0, longitude: 3.0}
+          ]
         }}
   """
   @spec route_segments(
           nonempty_list(Util.Location.From.t()),
           Util.Location.From.t(),
           Util.Location.From.t()
         ) :: {:ok, Result.t()} | :error
-  def route_segments(
-        [],
-        _connection_start,
-        _connection_end
-      ),
-      do: :error
+  def route_segments([], _connection_start, _connection_end), do: :error
+  def route_segments([_], _connection_start, _connection_end), do: :error
 
   def route_segments(
         shape,
         connection_start,
         connection_end
       ) do
-    {_, start_index} =
-      shape
-      |> Enum.with_index()
-      |> Enum.min_by(fn {point, _index} -> Util.Location.distance(point, connection_start) end)
+    {nearest_start_point, start_index} =
+      nearest_point_to_shape(shape, connection_start)
 
-    {_, end_index} =
-      shape
-      |> Enum.with_index()
-      |> Enum.min_by(fn {point, _index} -> Util.Location.distance(point, connection_end) end)
+    {nearest_end_point, end_index} =
+      nearest_point_to_shape(shape, connection_end)
 
     {:ok,
      %__MODULE__.Result{
-       before_detour: Enum.slice(shape, 0..start_index),
-       detour: Enum.slice(shape, start_index..end_index),
-       after_detour: Enum.slice(shape, end_index..-1)
+       before_detour: Enum.slice(shape, 0..start_index) ++ [nearest_start_point],
+       detour:
+         [nearest_start_point] ++
+           Enum.slice(shape, (start_index + 1)..end_index) ++ [nearest_end_point],
+       after_detour: [nearest_end_point] ++ Enum.slice(shape, (end_index + 1)..-1)
```

Comment:
FWIW, I spent a bit of time trying to figure out what I wanted from a "clean up" and I didn't come to any satisfying solutions.

The one thing that made it _marginally_ better, was pulling out the `slice` results into their own values before concatenating them, but I'm not convinced it's worth.

I also tried using `Enum.split` instead, since the current behaviour is a little wrong, but that didn't make anything cleaner [somewhat expected though] :/

Reply:
if I could have _anything_, I'd align these via newlines and have the `++` operator be at the beginning, but I'm pretty sure the formatter would blow these changes away.

```suggestion
       before_detour: 
         Enum.slice(shape, 0..start_index)
         ++ [nearest_start_point],
       detour:
         [nearest_start_point]
         ++ Enum.slice(shape, (start_index + 1)..end_index)
         ++ [nearest_end_point],
       after_detour:
       	[nearest_end_point]
       	++ Enum.slice(shape, (end_index + 1)..-1)
```

Reply:
The formatter _partly_ allows what you're looking for, but it shoved the `++`'s to the end of the previous line rather than the beginning of the next line.

I agree with you that having the `++` at the beginning of the next line, but I think breaking it into separate lines is an improvement over what's currently there, so I'm gonna roll with it.

Reply:
[Improvement!](https://github.com/mbta/skate/pull/2506/commits/c0da15e26097a634ee4cb5c29031b2eae49644ec)